swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt

```diff
@@ -149,39 +149,43 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
    * This function is used when a caller makes an opt out request through the controller
    *  and the resource has not already been marked.
    */
-  override fun optOut(resourceId: String, workConfiguration: WorkConfiguration) {
-    val resource = getCandidate(resourceId, "", workConfiguration)
-      ?: return
-    log.info("Opting out resource $resourceId in namespace ${workConfiguration.namespace}")
-    val resourceWithOwner = listOf(resource).withResolvedOwners(workConfiguration).first()
-    val newMarkedResource = MarkedResource(
-      resource = resourceWithOwner,
-      summaries = emptyList(),
-      namespace = workConfiguration.namespace,
-      resourceOwner = resourceWithOwner.details[resourceOwnerField] as String,
-      projectedDeletionStamp = deletionTimestamp(workConfiguration),
-      lastSeenInfo = resourceUseTrackingRepository.getLastSeenInfo(resourceWithOwner.resourceId)
-    )
+  override fun optOut(resourceId: String, workConfiguration: WorkConfiguration): ResourceState {
+    val resource = getCandidate(
+      resourceId = resourceId,
+      workConfiguration = workConfiguration
+    ) ?: throw NotFoundException()
 
-    val status = Status(Action.OPTOUT.name, clock.instant().toEpochMilli())
-    val state = ResourceState(
-      markedResource = newMarkedResource,
-      deleted = false,
-      optedOut = true,
-      statuses = mutableListOf(status),
-      currentStatus = status
-    )
+    val markedResource = resourceRepository.find(resourceId, workConfiguration.namespace)
+      ?: MarkedResource(
+        resource = resource,
+        summaries = emptyList(),
+        namespace = workConfiguration.namespace,
+        projectedDeletionStamp = -1L
+      )
 
-    // opt out here to ensure there is never a time when this resource is not opted out.
-    resourceStateRepository.upsert(state)
-    resourceRepository.upsert(newMarkedResource)
+    val status = Status(Action.OPTOUT.name, clock.instant().toEpochMilli())
+    val optOutState = resourceStateRepository.get(resource.resourceId, workConfiguration.namespace)
+      ?: ResourceState(
+          markedResource = markedResource,
+          deleted = false,
+          optedOut = true,
+          statuses = mutableListOf(status),
+          currentStatus = status
+        )
 
-    // this will trigger the actual opt out action
-    applicationEventPublisher.publishEvent(OptOutResourceEvent(newMarkedResource, workConfiguration))
+    log.info("Opting out resource ${markedResource.resourceId} in namespace ${workConfiguration.namespace}")
+    ensureResourceUnmarked(markedResource, workConfiguration, "Opted Out")
+    applicationEventPublisher.publishEvent(OptOutResourceEvent(markedResource, workConfiguration))
+    return optOutState.copy(
+      currentStatus = status,
+      statuses = setOf(optOutState.statuses + status).flatten().toMutableList()
+    ).apply {
+      resourceStateRepository.upsert(this)
```

Comment:
what about `resourceRepository.upsert()` ? will this take care of resource upsert ?

Reply:
Why would we do that while we are ensuring this resource is opted out & umarked?

Reply:
ok so we don't need to update the resource state , i ask this because the previous optout flow had this call?

Reply:
That was definitely a bug. resourceRepository.upsert would end up marking the resource. ResourceStateRepository.upsert does update the status to opted out which is what we need